ntoskrnl/se/accesschk.c

```diff
@@ -486,7 +424,8 @@ SepAccessCheck(
     BOOLEAN Defaulted;
     NTSTATUS Status;
     PACCESS_TOKEN Token = NULL;
-    PACCESS_CHECK_RIGHTS AccessCheckRights = NULL;
+    ACCESS_MASK AccessRightsArray[ACCESS_RIGHTS_MAX_ARRAY];
```

Comment:
Observation: This array contains only 3 (== ACCESS_RIGHTS_MAX_ARRAY) ACCESS_MASK elements, not more, not less.
So, you could have kept your original `ACCESS_CHECK_RIGHTS` structure, and here, declare a variable of type `ACCESS_CHECK_RIGHTS` (instead of that array, or a pointer as it was before). This would allow to get rid of the ugly indices define names from above too, and to keep the original code almost unchanged.
The only thing that would be removed is, the pool allocation + the pool free at the end.

Reply:
Ok, it turns out ditching the structure that I introduced is a bad idea. I'm going to refactor this as per your observation.